### DISCUSSION TITLE
Example for connecting to a LSP server on the web worker?

### DISCUSSION THREAD

**Discussion by ejfasting** [TypeFox/monaco-languageclient#597] [Q&A] [answered]
Hello, 

I have an existing vscode web extension that has a custom LSP server (which is running on the webworker). 
I'm trying to figure out how to create a client that can be used by the monaco editor, and I want it to be as basic as possible so that i can build it piece by piece. 

Optimally I would like to reuse the existing extension i have for vscode, and its a little confusing that the documentation for monaco-vscode-api says it supports vscode extensions (?). 
'You' also have an ongoing discussion (from what i can tell), so I'm a little unsure which NPM package to actually use:
https://github.com/TypeFox/monaco-languageclient/issues/543 

If anyone could point me in the right direction that would be great. As a code-example I've found this:
https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-web-extension-sample
This is an example-code for an LSP web extension for vscode, and basically i would like to add another client to this (without changing the LSP server). 

//e

**Comment by CGNonofr** [comment]
If you want to run your vscode extension as is, monaco-vscode-api can indeed do that. You can load your vsix using the provided vsix rollup loader. Depending on the feature you want, you may need to enable some services overrides.

**Reply by ejfasting** [reply]
Thank you so much for the reply! 

This pretty much confirms what I thought, which is very helpful :) 
Is there any reason to NOT do it this way? I'm not sure if there is a difference in functionality, or if everything is solveable by including the service overrides (as you say).

**Reply by CGNonofr** [reply]
If you include all service overrides and load your extension, you're supposed to have the exact same behavior as VScode itself. In your case, I don't see any reason to not implement it that way

monaco-languageclient/vscode/json-ws-jsonrpc are designed to run a language server on a remote server via websocket

**Reply by ejfasting** [reply]
I see, thank you! Then it would, in my case, make the most sense to use the vsix from the existing extension instead making a client 'only' for monaco. 
That way i only have to maintain 1 client :)

I havent found a good example on how to do it. Do you know of some code i can look at? 

Thanks, but in my case i want to have each user spin up their own lsp in their own browser.
The end result is gonna have thousands of simultaneous users, so i want to offload to each user instead of hosting an lsp on a server. 
But thank you for the input, much appreciated! 

/ej

**Reply by ejfasting** [reply]
@CGNonofr Sorry to bother you again.. I did some testing yesterday but was unable to make it work. Are you supposed to just do a simple "import './extension.vsix" at the top of for example main.ts? 

I'm also wondering what the difference is between a remoteExtension and loading the vsix (as you suggested)? 
I looked at this example: https://github.com/CodinGame/monaco-vscode-api/blob/main/demo/src/features/remoteExtension.ts , and it looks to just use the vscode api directly, and dont require you to bundle the vscode-extension into a vsix (?). 

//ej

**Reply by CGNonofr** [reply]
A vsix file is just a tarball. There is also a directory extension rollup plugin if you don't want to create a vsix. The remote extension uses the extension host server and run the extension in a node environment, not in the browser

**Reply by ejfasting** [reply]
@CGNonofr 
How would you go about using a vsix? I'm trying to figure out how it's done, as the documentation for how to import and use the vsix is not very descriptive:
https://github.com/CodinGame/monaco-vscode-api?tab=readme-ov-file#loading-vsix-file 
Those 4 lines does nothing when i run it locally, so I'm missing a piece.. 
I havent found any good examples either, which is unfortunate.

If the vsix only contains an lsp which returns some completionItems + a vscode extension that is the client. How would you import the vsix and have the monaco editor give you the same completionItems as you get in vscode? Is there a step-by-step i can follow?

**Reply by CGNonofr** [reply]
Importing a vsix using the vsix plugin will only do something if the extension is a web compatible extension (nothing magical here) an unfortunately it will just silently do nothing if it's not the case

if you want to run a non-web compatible extension, you'll need the to run the node extension host server on a server of yours

**Reply by ejfasting** [reply]
hm, it is a webextension. Actually package.json (in the extension) contains attributes for both main and browser, so that it builds (with esbuild) 2 different versions. 

So when this is built the extension should be webcompatible/a webextension. 
But what can i expect to happen if it gets imported 'successfully'? Will I have to write any code in the monaco-extension, or will it automatically add all the things the extension/vsix contains? Feels like the explanation in the link in my previous post is not sufficient to 'load things up' properly..

**Reply by CGNonofr** [reply]
It's supposed to do the exact same thing as installing the extension in VSCode, with the difference that the extension will be listed as `builtin` in the interface (if you use the extension gallery service override)

Do you mind sharing the vsix so I can have a look?

**Reply by ejfasting** [reply]
Sure, no problem :)
You can find the .vsix here: https://github.com/ejfasting/MonacoTest/blob/main/jsfso-0.0.1.vsix 
(nevermind the repo, its basically an attempt of going a different route and just import the languageServer directly, instead of through a vsix). 

Import that in vscode, open a plain text document and click ctrl+space, and the result should be this: 
![bilde](https://github.com/TypeFox/monaco-languageclient/assets/60758190/fa7c31e0-4dbf-4e32-aa60-fd2ab73e8178)

**Reply by CGNonofr** [reply]
Did you manage to make it work on the web version of VSCode?

**Reply by ejfasting** [reply]
Well, i assumed that running it locally with arg '--extensionDevelopmentKind=web' would simulate that it was working in vscode.dev. 
I will take it one step further and run it with [@vscode/test-web](https://github.com/microsoft/vscode-test-web) and/or try to sideload it instead. 

But, with this vsix working, how would a sample project look like? I'm still unsure if I'm actually loading it up as i should (just putting import at the top of main.ts). I will make a complete repo you can look at, if I'm unable to make it work after validating the .vsix

**Reply by CGNonofr** [reply, marked as answer]
I've just tried adding in in the monaco-vscode-api demo and It worked on the first try: 
![image](https://github.com/TypeFox/monaco-languageclient/assets/23257479/b9b17af3-e3be-4efd-bd4b-f4fb42b595ba)

Can you please open a discussion in monaco-vscode-api instead to not pollute to much here as it's not really related to monaco-languageclient

**Reply by ejfasting** [reply]
Roger that! 
I was able to validate the vsix (which you also did i see), so the issue is how i load it up in my code. I'm starting another thread on monaco-vscode-api and closing this one. Thanks!

**Comment by msujew** [comment]
In addition to what @CGNonofr said, there is an example in this repo for [webworker based language servers](https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/langium/statemachineClient.ts). In this case, the language server is provided by [Langium](https://github.com/eclipse-langium/langium).

**Reply by ejfasting** [reply]
Hello! 

I looked at the sample, and threw together a small repo with a very stripped example:
https://github.com/ejfasting/MonacoTest 

Would this be the correct approach? If not, could you help me out? :)

Thank you for the reply!
//Eivind

**Reply by msujew** [reply]
I believe you're still missing the language definition (otherwise, `languageId` has no actual language association), see:

https://github.com/TypeFox/monaco-languageclient/blob/b912f1212ac03cd53ba36d52a757bd1911ee0e82/packages/examples/src/langium/statemachineClient.ts#L55-L89

Also, otherwise this looks mostly alright. There might be a name mismatch in the language server (the file is named `lsp-server.js`, while the worker requests `server.js`), but that depends on how you serve it.

**Reply by ejfasting** [reply]
Thank you! :) 

I will keep working from this then. Actually decided to use monaco-vscode-api, and import the vsix instead, but this is great to get more familiar with how this works. 

Again, thanks a lot for the reply! 

//eivind

**Reply by ejfasting** [reply]
@msujew Hi again!
Sorry, I had to put this to the side and focus on other things for a while.. 

Anyways, i tried to add the language definition, and I've fixed the typo. 
I also added Vite, to make it easier to develop/work with. 

It looks like the server is being loaded up, as I've added a console.log in the browser.js that is posted in the console. 
But it gives no code completions, which is basically all the server now should return. 

It also complains about vscode/localExtensionHost, but if i add that and await initServices() it never gets past that point.. 

![bilde](https://github.com/TypeFox/monaco-languageclient/assets/60758190/52314cfb-0784-4056-905a-cbdaa5723e53)

Would it be possible for you to have another look, and possibly help out figure out what I'm doing wrong? I probably miss a piece here, but I have no idea what (been going through the langium-example a dozen times without figuring it out).

//Eivind

**Comment by kaisalmen** [comment]
@ejfasting as @msujew already said this looks mostly good.
We have also created `monaco-editor-wrapper` (https://github.com/TypeFox/monaco-components) that allows to reduce the boiler plate code needed to build a client for a worker or other language servers. There are a couple of example in that repo as well that show the usage from simple to more advanced.
